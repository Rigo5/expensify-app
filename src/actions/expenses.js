import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';

//EXPENSES ACTIONS

//adesso che dobbiamo integrare firebase andremo a fare il dispatch di funzioni che verranno eseguite da Redux
export const addExpense = (expense) =>({
    type: 'ADD_EXPENSE',
    expense
});

//La funzione che viene tornata viene eseguita da redux il quale gli passerÃ  come argomento i dispatch
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt }
        return database.ref('expenses').push(expense).then( (ref) => {
            dispatch(addExpense({
                id : ref.key,
                ...expense
            }))
        })
    };
}

export const removeExpense = ({ id = null} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({ id = null} = {}) => {
    return (dispatch) => {
        return database.ref('expenses/' + id).set(null).then( () => {
            console.log('rimosso'); 
            dispatch(removeExpense({id : id}));
        })
    }
}

export const editExpense = (id, updates) => ({
    type : 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpenses = (id, updates) => {
    return (dispatch) => {
        return database.ref('expenses/' + id).update({
            ...updates
        }).then( () => {
            dispatch(editExpense(id, updates))
        })
    }
}

//SET EXPENSES: prende array setta come stato
export const setExpenses = (expenses) => ({
    type : 'SET_EXPENSES',
    expenses
})

//start set Expense : prende dati da db e dispatcha set Expenses 
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((snapshotChild) => { 
                expenses.push({
                    id : snapshotChild.key,
                    ...snapshotChild.val()
                });
            });
            console.log(expenses)
            dispatch(setExpenses(expenses));
        })
    }
}
