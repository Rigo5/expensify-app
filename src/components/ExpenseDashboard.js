import React from 'react';
import ExpenseList from './expensesList';
import ExpenseListFilter from './ExpensesListFilter';

const ExpenseDashboard = () => {
    return (
        <div>
            <p>la merda di cane puzza</p>
            <ExpenseListFilter />
            <ExpenseList />
        </div>
    );
}

export default ExpenseDashboard;