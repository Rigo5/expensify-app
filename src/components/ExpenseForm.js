import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



class ExpenseForm extends React.Component {
    
    constructor(props){
        super(props); 
        console.log(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : "Notes which are optional",
            amount: props.expense ? props.expense.amount.toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState((prevState) => {
            return {
                description: description
            }
        })
    }

    onNoteChange = (e) => {
        //se non la metto in una variabile e lo uso direttamente nel set state va in erroe e devo usare e.persist()
        const note = e.target.value;
        this.setState((prevState) => ({
            note: note
        }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        //se matcia formato 122331.12 viene aggiornato altrmimenti sticazzi
        if (amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount: amount }))
        }
    }

    onDateChange = (createAt) => {
        if (createAt) {
            this.setState(() => ({
                createdAt: createAt
            }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }))
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            //please description and amount 
            this.setState(() => ({
                error: 'Please set description and amount'
            }))
        } else {
            if (this.state.error) {
                this.setState(() => ({
                    error: ''
                }))
            }
            console.log(this.state.createdAt.valueOf());
            //clear the error
            this.props.onSubmit({
                description : this.state.description,
                amount : parseFloat(this.state.amount),
                createdAt : this.state.createdAt.valueOf(),
                note : this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error ? <p>Fuck you self</p> : null}
                <form onSubmit={this.onSubmit}>
                    <input type="text"
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoFocus />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange} />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea placeholder={this.state.note} onChange={this.onNoteChange} />
                    {this.state.amount ? <button>Edit Expense</button>: <button>Create Expense</button>}
                </form>
            </div>
        );
    }
}

export default ExpenseForm;