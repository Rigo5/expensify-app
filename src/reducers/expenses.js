

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //this is callled sÃ¨read operator
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id != action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id === action.id ? {...expense, ...action.updates} : expense);
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
}

export default expensesReducer;