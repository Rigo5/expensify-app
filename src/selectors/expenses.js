//SELECTORS: Sono funzioni per recuperare dati dai container redux
//In questo caso recuperiamo le spese utilizzando tutti i filtri possibili

//Ora utlizziamo il filtro per effettivamente filtrare le spese creiamo una funzione che faccai questo
//che mettiamo in subscribe cosi ogni volta che cambiamo qualcosa sbombiamo il tutto
import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate}) => {

    return expenses.filter( (expense) =>
        {
            const createdAtMoment = moment(expense.createdAt);
            const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
            const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
            const textMatch = expense.description.toLowerCase().includes(text); 
            return startDateMatch &&  endDateMatch &&  textMatch;
        }
    ).sort((a, b) => {
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        if(sortBy === 'amount'){
            return a.amount < b.amount ? -1 : 1;
        }
    });
}

export default getVisibleExpenses;