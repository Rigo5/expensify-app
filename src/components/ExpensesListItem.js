import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';

export const ExpensesListItem = (props) => (
    <div>
        <Link to={'/edit/' + props.id}>{props.description}</Link>
        <p>{props.amount} - {props.createdAt}</p>
        <button onClick={(e) => {
            props.onClick({ id: props.id });
        }}>Remove</button>
    </div>
)

const mapDispatchToProps = (dispatch) => ({
    onClick : (id) => removeExpense({ id : id})
});


export default connect(undefined, mapDispatchToProps)(ExpensesListItem);