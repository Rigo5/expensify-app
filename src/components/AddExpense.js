import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense, editExpense } from '../actions/expenses';


/* un problema qua per il testing è dato da : 
    props.dispatch(addExpense({...expense}))
    props dispatch sarà messa come spy, il problema è che come arg
    ha una referenza a una funzione di un modulo e non è passata come prop
    redux per fortuna ci viene in contro e ci permette di definire come per lo stato
    una funzione che mappa i vari dispatch come props
    il codice di prima diventa props.onSubmit(expense); che è più testabile
    in qunato abbiamo astratto il dispatch. 
    Per evitare inline functions sarebbe meglio portare stateless function a componente
    FAAREMO 
    */
export const AddExpense = (props) => {
    return (
        <div>
            <p>aggiungi spese</p>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.onSubmit(expense);
                    props.history.push('/');
                }}

            />
        </div>
    );
}

const mapDispatchToProps = (dispatch) => ({
    onSubmit : (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpense);