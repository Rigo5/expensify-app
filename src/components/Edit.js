import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense } from '../actions/expenses';



export class Edit extends React.Component {


    onSubmit = (expense) => {
        console.log('edited');
        this.props.onSubmit(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <p>Editing l'oggetto: {this.props.expense.id}</p>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
            </div>);
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch) => ({
    onSubmit : (id, expense) => dispatch(editExpense(id, expense))
})

export default connect(mapStateToProps, mapDispatchToProps)(Edit);