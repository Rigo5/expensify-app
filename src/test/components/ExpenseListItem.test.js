import React from 'react'; 
import { shallow } from 'enzyme';
import { ExpensesListItem } from '../../components/ExpensesListItem';
import { expenses } from '../../fixtures/expenses';


let wrapper, onClick; 
beforeEach(() => {
    onClick = jest.fn(); 
    wrapper = shallow(<ExpensesListItem key={expenses[0].id} {...expenses[0]} onClick={onClick} />);
});

test('shoul list item render with one expense', () => {

    expect(wrapper).toMatchSnapshot();
})


test('should remove handler', () => {

    wrapper.find('button').simulate('click');

    expect(onClick).lastCalledWith(expenses[0].id);
})