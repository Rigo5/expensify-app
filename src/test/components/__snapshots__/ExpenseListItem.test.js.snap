// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shoul list item render with one expense 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    merda
  </Link>
  <p>
    10
     - 
    0
  </p>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
